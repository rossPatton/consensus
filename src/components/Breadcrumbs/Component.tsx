import React, { memo } from 'react';
import { Link } from 'react-router-dom';

export const BreadcrumbsComponent = memo((props: any) => (
  <ul className="fs6 fw600 mB2 lh1 lsNone fx fxWrap brdB1 mB3 pB3">
    {console.log('props => ', props)}
    {props.country && (
      <li className="mR1">
        {!props.region && props.country.name}
        {props.region && (
          <Link to={`/${props.country.code}`}>
            {props.country.name}
          </Link>
        )}
      </li>
    )}
    {props.region && (
      <li className="mHide mR1">/</li>
    )}
    {props.region && (
      <li className="mR1">
        {!props.city && props.country.name}
        {props.city && (
          <Link to={`/${props.country.code}/${props.region.code}`}>
            {props.region.name}
          </Link>
        )}
      </li>
    )}
    <li className="mHide mR1">/</li>
    <li className="mR1">
      <Link to="/us/ny/new-york-city">
        New York City
      </Link>
    </li>
  </ul>
));
