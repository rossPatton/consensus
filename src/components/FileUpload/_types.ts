export type tProps = {
  // what type of file are we uploading
  fieldKey: string,
  // where in DO spaces to upload the file
  folder?: 'groups' | 'users',
  // the unique identifier for the image
  hash?: string,
  img?: string,
  // if we want to include additional info about uploading the image to the user, like size
  info?: string,
  // shorthand code that includes group or user id and image type
  prefix?: string,
  // header text for the section
  title?: string,
  // just used here for rendering the preview
  width?: string,
}

export type tContainerProps = tProps & {
  featuredImage: ts.thunk<string>,
  postFeaturedImageDispatch: (query: ts.spacesQuery) => void,
}

export type tComponentProps = tProps & {
  removeImage: (ev: React.MouseEvent<HTMLButtonElement>) => void,
  setImage: (ev: React.ChangeEvent<HTMLInputElement>) => void,
};
