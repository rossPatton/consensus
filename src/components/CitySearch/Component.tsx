import _ from 'lodash';
import React, {memo} from 'react';

import stateCodeMap from '../../json/usa/stateCodeMap.json';
import {tComponentProps} from './_types';

export const CitySearchComponent = memo((props: tComponentProps) => {
  const onCityChange = (ev: React.ChangeEvent<HTMLSelectElement>) => {
    const city = _.find(props.cities, c => c.name === ev.currentTarget.value);
    if (city) {
      props.updateState({
        city: city.name,
        cityId: city.id,
        region: city.region,
        regionId: city.regionId,
      });
    }
  };

  const onRegionChange = (ev: React.ChangeEvent<HTMLSelectElement>) => {
    const region = ev.currentTarget.value;
    if (region !== props.region) {
      props.updateState({
        city: '',
        cityId: 0,
        region,
        regionId: 0,
      });
    }
  };

  return (
    <>
      {props.showRegionField && (
        <>
          <h2 className="fs5 mB1 lh1">
            Pick a different state
          </h2>
          <label htmlFor="stateSelect">
            <select
              name="type"
              id="stateSelect"
              className="mB4 row bgWhite"
              value={props.region}
              onBlur={onRegionChange}
              onChange={onRegionChange}>
              {Object.keys(stateCodeMap).map(state => (
                <option key={state} value={state}>
                  {state}
                </option>
              ))}
            </select>
          </label>
        </>
      )}
      <h2 className="fs5 mB1 lh1">
        City in <span className="dInBl mR2">
          {(!props.showRegionField && props.geo.region) && props.geo.region}
          {(props.showRegionField && props.region) && props.region}
        </span>
        {!props.showRegionField && (
          <button
            type="button"
            onClick={props.toggleRegionField}>
            Not the right state?
          </button>
        )}
      </h2>
      {props.label && (
        <p className="fs5 copyBlack mB1">
          {props.label}
        </p>
      )}
      <div className="mB4">
        {props.cityId === 0 && (
          <label htmlFor="citySelect">
            <select
              name="type"
              id="citySelect"
              className="row bgWhite"
              value={props.city}
              onBlur={onCityChange}
              onChange={onCityChange}>
              {props.cities.map((city: tCity) => (
                <option key={city.id} value={city.name}>
                  {city.name}, {city.region}
                </option>
              ))}
            </select>
          </label>
        )}
        {props.cityId > 0 && (
          <div className="brdA1 p3 black br8 dInBl">
            <b>{props.city}</b>, <span className="dInBl mR3">{props.region}</span>
            <button
              type="button"
              onClick={() => {
                props.updateState({
                  city: '',
                  cityId: 0,
                });
              }}>
              Not the right city?
            </button>
          </div>
        )}
      </div>
    </>
  );
});
