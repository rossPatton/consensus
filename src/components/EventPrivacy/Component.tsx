import React, {memo} from 'react';

import {tProps} from './_types';

export const EventPrivacyComponent = memo((props: tProps) => (
  <small className="bgYellowLite br8 p1 pL2 pR2 mR2">
    {props.isPrivate && (
      <>
        <span
          role="img"
          className="mR1"
          aria-label="Lock Emoji">
          ðŸ”’
        </span>
        Private Event
      </>
    )}
    {!props.isPrivate && (
      <>
        <span
          role="img"
          className="mR1"
          aria-label="Tada Emoji">
          ðŸŽ‰
        </span>
        Public Event
      </>
    )}
  </small>
));
