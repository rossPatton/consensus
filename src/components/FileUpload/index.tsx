import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import {v4} from 'uuid';

import {postUpload} from '~app/redux';
import {success as postUploadSuccess} from '~app/redux/uploads/post/actions';

import {tContainerProps, tProps, tStore} from './_types';
import {FileUploadComponent} from './Component';

/**
 * @description A Component for uploading files to the consensus DO spaces account
 * must be used as part of a form with multi-part encoding
 */
class FileUploadContainer extends React.Component<tContainerProps> {
  static defaultProps = {
    folder: 'groups' as 'groups' | 'users',
    hash: v4(),
    img: '',
    info: 'We recommend a size of at least 760x428px',
    prefix: '',
    title: 'Upload Image',
    width: '60',
  };

  removeImage = (ev: React.MouseEvent<HTMLButtonElement>) => {
    ev.preventDefault();
    const upload = {[this.props.fieldKey]: ''};
    const fileInput = document.getElementById('fileUpload') as HTMLInputElement;
    fileInput.value = '';
    this.props.dispatch(postUploadSuccess(upload));
  }

  setImage = async (ev: React.ChangeEvent<HTMLInputElement>) => {
    ev.preventDefault();

    const regexExt = /[^\\]*\.(\w+)$/;
    let image = `${this.props.prefix}:${this.props.hash}`;
    let ext = 'jpg';

    const body = new FormData();
    const fileInput = document.getElementById('fileUpload') as HTMLInputElement;
    if (fileInput !== null) {
      const { files }: { files: FileList | null } = fileInput;
      if (files !== null) {
        ext = files[0].name.match(regexExt)[1];
        image = `${image}.${ext}`;
        body.append(this.props.fieldKey, files[0], image);
      }
    }

    return this.props.postUploadDispatch(body);
  }

  render() {
    return (
      <FileUploadComponent
        fieldKey={this.props.fieldKey}
        folder={this.props.folder}
        img={this.props.img}
        info={this.props.info}
        title={this.props.title}
        removeImage={this.removeImage}
        setImage={this.setImage}
        width={this.props.width}
      />
    );
  }
}

const mapStateToProps = (store: tStore, props: tProps) => {
  const img = store.uploads.data[props.fieldKey as string];
  return { img };
};

const mapDispatchToProps = (dispatch: Function) => ({
  dispatch,
  postUploadDispatch: (query: ts.spacesQuery) => dispatch(postUpload(query)),
});

// @ts-ignore
export default connect(mapStateToProps, mapDispatchToProps)(FileUploadContainer);
