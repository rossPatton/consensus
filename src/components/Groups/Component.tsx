import cx from 'classnames';
import _ from 'lodash';
import React, {memo} from 'react';

import {Group} from './_components';
import {tComponentProps} from './_types';

export const GroupsComponent = memo((props: tComponentProps) => (
  <ul className={cx({'flex flex-wrap': !props.asList})}>
    {props.groups.length > 0
      && props.pendingOrgs.length > 0
      && (
        <li className="font-bold mb-2 pb-2 border-b-1">
          Current group memberships
        </li>
      )}
    {props.groups.map((group, i) => (
      <Group
        {...props}
        key={group.handle}
        group={group}
        index={i}
      />
    ))}
    {props.pendingOrgs.length > 0 && (
      <>
        <li className="font-bold mb-2 pb-2 border-b-1">
          Pending group memberships
        </li>
        {props.pendingOrgs.map((pendingOrg, i) => (
          <Group
            {...props}
            index={i}
            key={pendingOrg.id}
            group={pendingOrg}
          />
        ))}
      </>
    )}
  </ul>
));
