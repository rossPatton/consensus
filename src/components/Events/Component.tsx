import cx from 'classnames';
import dayJS from 'dayjs';
import _ from 'lodash';
import qs from 'query-string';
import React, {memo} from 'react';
import {Link} from 'react-router-dom';

import {ExternalLink, PlaceholderImage, RSVP} from '../../components';
import {objToQueryString} from '../../utils';
import {tComponentProps} from './_types';

export const EventsComponent = memo((props: tComponentProps) => (
  <ul
    className={cx({
      'fx fxdRow taL': props.horizontal,
      'jcBetween': props.horizontal && props.events.length === 4,
    })}>
    {props.events.map((ev, i) => {
      const isPastEvent = dayJS(ev.date).isBefore(dayJS());

      return (
        <li
          key={ev.id}
          className={cx({
            'mB2': !props.horizontal,
            'col row': props.horizontal,
            'mR4': props.horizontal && i !== props.events.length - 1,
          })}>
          <div
            className={cx({
              'fx aiCtr': !props.horizontal,
              'p3 hvrBgGrey1 br8': props.isEditable,
            })}>
            <div
              className={cx({
                'mB2 bgWhite': props.horizontal,
                'mR3 bgGrey2': !props.horizontal,
              })}>
              <Link
                className="noUnderline"
                to={ev.isDraft
                  ? `/org/${ev.orgId}/planMeeting?${qs.stringify(ev)}`
                  : `/event/${ev.id}`}>
                <PlaceholderImage
                  height={100}
                  seed={ev.id}
                  width={200}
                />
              </Link>
            </div>
            <div>
              <div
                className={cx({
                  'fx aiCtr mB1 fs7 fw600 lh1': true,
                  o5: isPastEvent,
                })}>
                <time className="mR1" dateTime={ev.date}>
                  {dayJS(ev.date).format('MMM DD | h:mmA')}
                </time>
                {!props.horizontal && (
                  <>
                    <span className="mR1">@</span>
                    {ev.locationLink && (
                      <ExternalLink
                        noFollow
                        className="mR1"
                        to={ev.locationLink}>
                        {ev.location}
                      </ExternalLink>
                    )}
                  </>
                )}
                {!ev.locationLink && ev.location}
              </div>
              <h3
                className={cx({
                  'fx aiCtr ttCap': true,
                  mB1: props.isEditable,
                  fs4: props.horizontal,
                })}>
                <Link
                  className="noUnderline"
                  to={ev.isDraft
                    ? `/org/${ev.orgId}/planMeeting?${qs.stringify(ev)}`
                    : `/event/${ev.id}`}>
                  {ev.title}
                </Link>
              </h3>
              {props.showRSVPs && (
                <div
                  className={cx({
                    'fx aiCtr fs6 lh1': true,
                    hide: props.horizontal,
                    o5: isPastEvent,
                  })}>
                  {!ev.isDraft && <RSVP event={ev} />}
                  {ev.isDraft && 'This event is not published yet. Click to edit.'}
                </div>
              )}
              {props.showOrgName
                && (
                  <Link
                    to={`/org/${ev.orgId}`}
                    className={cx({
                      'fw600 fs6 lh1': true,
                      o5: isPastEvent,
                    })}>
                    {ev.orgName}
                  </Link>
                )}
              {!isPastEvent
              && props.isEditable
              && (
                <div className="fx aiCtr mB1">
                  <div className="bgGrey4 br4 fs7 lh1 mR2 p1 pL2 pR2 white">
                    {ev.isDraft ? 'Draft' : 'Published'}
                  </div>
                  <div className="fx aiCtr mR1">
                    <Link
                      to={`/org/${ev.orgId}/planMeeting?${objToQueryString(ev)}`}
                      className="btn fs7 fw600 hvrBgGrey1 lh1 noUnderline p1 pL2 pR2">
                      <span
                        role="img"
                        aria-label="Hand with Pen Emoji">
                        ✍️
                      </span>
                      Edit
                    </Link>
                  </div>
                  <button
                    onClick={e => props.deleteEvent(e, ev.id)}
                    className="aiCtr fs7 fw600 fx hvrBgGrey1 lh1">
                    <span
                      role="img"
                      className="mR1"
                      aria-label="Big X Emoji">
                      ✖️
                    </span>
                    Delete
                  </button>
                </div>
              )}
            </div>
          </div>
        </li>
      );
    })}
  </ul>
));
