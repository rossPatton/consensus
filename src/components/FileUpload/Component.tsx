import cx from 'classnames';
import React, {memo} from 'react';
import {v1} from 'uuid';

import {Emoji} from '~app/components';

import {tComponentProps} from './_types';

const baseUrl = 'https://consensus.nyc3.digitaloceanspaces.com';

export const FileUploadComponent = memo((props: tComponentProps) => (
  <>
    <h3 className="text-base mb-1">
      {props.title}
    </h3>
    <div
      id="imagePreview"
      className={cx({
        'animated bg-gray-1 p-3 mb-3 flex flex-col items-center': true,
        'fadeInDown': props.img,
        'hidden': !props.img,
      })}>
      <button
        type="button"
        onClick={props.removeImage}
        className="mb-2 p-2">
        <Emoji
          label="Close Emoji"
          emoji="✖️"
        />
        Remove Image
      </button>
      {props.img && (
        <img
          alt="Your uploaded file preview."
          height={props.width}
          width={props.width}
          src={`${baseUrl}/${props.folder}/${props.img}${__DEV__ ? `?v=${v1()}` : ''}`}
          className={cx({
            'rounded w-full': !props.fieldKey.includes('Avatar'),
            'rounded-circ': props.fieldKey.includes('Avatar'),
          })}
        />
      )}
    </div>
    <label
      htmlFor="fileUpload"
      className={cx({
        'fadeInDown bg-gray-1 border cursor-ptr flex flex-col items-center mb-3 p-3 rounded animated': true,
        'fadeInDown': !props.img,
        'hidden': props.img,
      })}>
      <input
        accept="image/png, image/jpeg, image/gif"
        id="fileUpload"
        name="eventFeaturedImage"
        onChange={props.setImage}
        type="file"
        className={cx({
          hidden: true,
        })}
      />
      <span className="btn flex items-center mb-1 p-1">
        <Emoji
          label="Picture (Camera) Emoji"
          emoji="📷"
        />
        Upload Your Image
      </span>
      {props.info && (
        <small>{props.info}</small>
      )}
    </label>
  </>
));
