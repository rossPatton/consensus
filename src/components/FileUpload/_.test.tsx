import React from 'react';
import { Provider } from 'react-redux';
import {create} from 'react-test-renderer';
import configureStore from 'redux-mock-store';

const mockStore = configureStore();

import FileUpload from '.';
import { FileUploadComponent } from './Component';

describe('components/FileUpload', () => {
  let store = mockStore({});

  beforeAll(() => {
    store = mockStore({
      uploads: {
        error: null,
        isLoading: false,
        data: {
          userAvatar: 'hashorsomething',
        },
      },
    });
  });

  it('container renders with cust img hash', () => {
    const component = create((
      <Provider store={store}>
        <FileUpload fieldKey="userAvatar" img="customHash" />
      </Provider>
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('container renders', () => {
    const component = create((
      <Provider store={store}>
        <FileUpload fieldKey="userAvatar" />
      </Provider>
    ));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('component renders', () => {
    const component = create((
      <Provider store={store}>
        <FileUploadComponent
          fieldKey="userAvatar"
          removeImage={jest.fn()}
          setImage={jest.fn()}
        />
      </Provider>
    ));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('component removeImage', () => {
    const component = create((
      <Provider store={store}>
        <FileUploadComponent
          fieldKey="userAvatar"
          removeImage={jest.fn()}
          setImage={jest.fn()}
        />
      </Provider>
    ));

    component.root.props.children.props.removeImage();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
