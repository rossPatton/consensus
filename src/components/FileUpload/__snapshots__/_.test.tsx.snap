// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FileUpload component removeImage 1`] = `
Array [
  <h3
    className="text-base mb-1"
  />,
  <div
    className="animated bg-gray-1 p-3 mb-3 flex flex-col items-center hidden"
    id="imagePreview"
  >
    <button
      className="mb-2 p-2"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      type="button"
    >
      Remove Image
    </button>
  </div>,
  <label
    className="fadeInDown bg-gray-1 border cursor-ptr flex flex-col items-center mb-3 p-3 animated fadeInDown"
    htmlFor="fileUpload"
  >
    <input
      accept="image/png, image/jpeg, image/gif"
      className="hidden"
      id="fileUpload"
      name="eventFeaturedImage"
      onChange={[MockFunction]}
      type="file"
    />
    <span
      className="btn flex items-center mb-1 p-1"
    >
      Upload your image
    </span>
  </label>,
]
`;

exports[`components/FileUpload component renders 1`] = `
Array [
  <h3
    className="text-base mb-1"
  />,
  <div
    className="animated bg-gray-1 p-3 mb-3 flex flex-col items-center hidden"
    id="imagePreview"
  >
    <button
      className="mb-2 p-2"
      onClick={[MockFunction]}
      type="button"
    >
      Remove Image
    </button>
  </div>,
  <label
    className="fadeInDown bg-gray-1 border cursor-ptr flex flex-col items-center mb-3 p-3 animated fadeInDown"
    htmlFor="fileUpload"
  >
    <input
      accept="image/png, image/jpeg, image/gif"
      className="hidden"
      id="fileUpload"
      name="eventFeaturedImage"
      onChange={[MockFunction]}
      type="file"
    />
    <span
      className="btn flex items-center mb-1 p-1"
    >
      Upload your image
    </span>
  </label>,
]
`;

exports[`components/FileUpload container renders 1`] = `
Array [
  <h3
    className="text-base mb-1"
  >
    Upload Image
  </h3>,
  <div
    className="animated bg-gray-1 p-3 mb-3 flex flex-col items-center fadeInDown"
    id="imagePreview"
  >
    <button
      className="mb-2 p-2"
      onClick={[Function]}
      type="button"
    >
      Remove Image
    </button>
    <img
      alt="Your uploaded file preview."
      className="rounded"
      src="https://consensus.nyc3.digitaloceanspaces.com/groups/hashorsomething"
      width="60"
    />
  </div>,
  <label
    className="fadeInDown bg-gray-1 border cursor-ptr flex flex-col items-center mb-3 p-3 animated hidden"
    htmlFor="fileUpload"
  >
    <input
      accept="image/png, image/jpeg, image/gif"
      className="hidden"
      id="fileUpload"
      name="eventFeaturedImage"
      onChange={[Function]}
      type="file"
    />
    <span
      className="btn flex items-center mb-1 p-1"
    >
      Upload your image
    </span>
    <small>
      We recommend a size of at least 760x428px
    </small>
  </label>,
]
`;

exports[`components/FileUpload container renders with cust img hash 1`] = `
Array [
  <h3
    className="text-base mb-1"
  >
    Upload Image
  </h3>,
  <div
    className="animated bg-gray-1 p-3 mb-3 flex flex-col items-center fadeInDown"
    id="imagePreview"
  >
    <button
      className="mb-2 p-2"
      onClick={[Function]}
      type="button"
    >
      Remove Image
    </button>
    <img
      alt="Your uploaded file preview."
      className="rounded"
      src="https://consensus.nyc3.digitaloceanspaces.com/groups/customHash"
      width="60"
    />
  </div>,
  <label
    className="fadeInDown bg-gray-1 border cursor-ptr flex flex-col items-center mb-3 p-3 animated hidden"
    htmlFor="fileUpload"
  >
    <input
      accept="image/png, image/jpeg, image/gif"
      className="hidden"
      id="fileUpload"
      name="eventFeaturedImage"
      onChange={[Function]}
      type="file"
    />
    <span
      className="btn flex items-center mb-1 p-1"
    >
      Upload your image
    </span>
    <small>
      We recommend a size of at least 760x428px
    </small>
  </label>,
]
`;
